lib/spack/spack/concretize.py

```diff
@@ -465,6 +472,54 @@ def _compiler_concretization_failure(compiler_spec, arch):
         raise UnavailableCompilerVersionError(compiler_spec, arch)
 
 
+def concretize_specs_together(*abstract_specs):
+    """Given a number of specs as input, tries to concretize them together.
+
+    Args:
+        *abstract_specs: abstract specs to be concretized
+
+    Returns:
+        List of concretized specs
+    """
+    # Create a temporary repository with an umbrella package
+    # that depends on each spec
+    tmpdir = tempfile.mkdtemp()
+    repo_path, namespace = spack.repo.create_repo(tmpdir)
+
+    debug_msg = '[CONCRETIZATION]: Creating helper repository in {0}'
+    tty.debug(debug_msg.format(repo_path))
+    pkg_dir = os.path.join(repo_path, 'packages', 'concretizationroot')
+    fs.mkdirp(pkg_dir)
+
+    environment = spack.tengine.make_environment()
+    template = environment.get_template('misc/coconcretization.pyt')
+
+    # Split recursive specs, as it seems the concretizer has issue
+    # respecting conditions on dependents expressed like
+    # depends_on('foo ^bar@1.0')
+    split_specs = []
+    for spec in abstract_specs:
+        split_specs.extend([s.strip() for s in spec.split('^')])
+
+    abstract_specs = [spack.spec.Spec(s) for s in abstract_specs]
+
+    with open(os.path.join(pkg_dir, 'package.py'), 'w') as f:
+        f.write(template.render(specs=[str(s) for s in split_specs]))
+
+    # Temporary use this repository on top of the existing ones
+    repo = spack.repo.Repo(repo_path)
+    with spack.repo.additional_repository(repo):
+        # Create a spec from that umbrella package and concretize it
+        concretization_root = spack.spec.Spec('concretizationroot')
+        concretization_root.concretize()
+        # Retrieve the direct dependencies
+        concrete_specs = [
+            concretization_root[spec.name] for spec in abstract_specs
```

Comment:
I think we need to copy the specs here. I don't think we want the resulting specs to refer into each other directly, just be guaranteed to refer to the same spec.

Reply:
Fair point. We probably don't want either to have the fake package listed into the dependents.